text/0000-interior.md

```diff
@@ -0,0 +1,348 @@
+embedded fields get dropped. These approaches will be discussed in the
+`Alternatives` section, below.
+
+# Detailed design
+
+We define an `Interior<T>` type and associated `"interior"` language
+item:
+
+```rust
+#[lang="interior"]
+struct Interior<T>(T);
+```
+
+An instance of `Interior<T>` is defined to be structurally identical
+to `T`, but is an independent type. Since it is an independent type,
+it can have a unique set of `impl`s defined against it: the `impl`s
+defined against `T` are not inherited by `Interior<T>`. In particular,
+this means that `Interior<T>` does not implement `Drop`, which means
+it is (by default) valid to partially or fully deconstruct an
+`Interior<T>` instance (even where it would not be valid to
+deconstruct an instance of `T`).
+
+We also define a new `DropValue` trait to replace the legacy `Drop`
+trait, and which takes an instance of `Interior<Self>` by value:
+
+```rust
+#[lang="dropvalue"]
+trait DropValue {
+    fn drop_value(Interior<Self>);
+}
+```
+
+Define the following bare functions to convert between `T` and
+`Interior<T>`:
+
+```rust
+fn into_interior<T>(t: T) -> Interior<T>;
+fn from_interior<T>(t: Interior<T>) -> T;
+fn as_interior<T>(t: &T) -> &Interior<T>;
+fn as_interior_mut<T>(t: &mut T) -> &mut Interior<T>;
+fn of_interior<T>(t: &Interior<T>) -> &T;
+fn of_interior_mut<T>(t: &mut Interior<T>) -> &mut T;
+```
+
+And then it is possible to define a compatibility `Drop` trait in
+terms of `DropValue`:
+
+```rust
+impl<T: Drop> DropValue for T {
+  fn drop_value(this: Interior<Self>) {
+    of_interior_mut(this).drop();
+  }
+}
+```
+
+## Compiler-generated drop-glue.
+
+In this design, the compiler-generated drop-glue for a type gets so
+simplified as to become nearly, or possibly entirely, non-existent.
+Consider the following types:
+
+```rust
+struct Foo;
+impl Drop for Foo { ... }
+
+struct Bar(Foo, Foo);
+impl Drop for Bar { ... }
+
+enum Baz {
+  Empty,
+  AmFoo(Foo),
+  AmBar(Bar),
+}
+impl Drop for Baz { ... }
+```
+
+In current Rust, the drop-glue for these types will look something
+like this:
+
+```rust
+fn drop_glue_Foo(v: Foo) {
+  v.drop();
+  mem::forget(v);
+}
+fn drop_glue_Bar(v: Bar) {
+  v.drop();
+  drop_glue_Foo(v.0);
+  drop_glue_Foo(v.1);
+  mem::forget(v);
+}
+fn drop_glue_Baz(v: Baz) {
+  v.drop();
+  match v {
+    Empty => (),
+    AmFoo(val) => drop_glue_Foo(val),
+    AmBar(val) => drop_glue_Bar(val),
+  }
+  mem::forget(v);
+}
+```
+
+Under this proposal, the drop-glue for member fields is called at
+the point `this` is consumed:
+
+```rust
+impl DropValue for Foo {
+  fn drop_value(this: Interior<Foo>) {
+    println!("drop_foo");
+    // no compiler-inserted clean-up necessary
+  }
+}
+impl DropValue for Bar {
+  fn drop_value(this: Interior<Bar>) {
+    println!("drop_bar");
+    // "this" is consumed at function exit, so the compiler
+    // will generate the following code:
+    // DropValue::<Foo>::drop_value(into_interior(this.0));
```

Comment:
Is it possible to do partial moves like this? Otherwise do:

```
let Bar(x, y) = this;
DropValue::<Foo>::drop_value(into_interior(x));
DropValue::<Foo>::drop_value(into_interior(y));
```